src/irmin-chunk/irmin_chunk.ml

```diff
@@ -74,40 +75,58 @@ module Chunk (K: Irmin.Hash.S) = struct
   let extract_data src srcoff dst dstoff len =
     Cstruct.blit src (offset_payload + srcoff) dst dstoff len
 
-  let type_of_int = function
-    | Indirect -> 0
-    | Data     -> 1
+  let type_indirect = 0
+  let type_data = 1
 
   let int_of_type = function
-    | 0 -> Indirect
-    | 1 -> Data
-    | _ -> failwith "Unknow type"
+    | Indirect -> type_indirect
+    | Data     -> type_data
 
-  let type_of_chunk x = int_of_type (get_type x)
+  (*let valid_type = function
+    | 0 | 1 -> ()
+    | _ -> failwith "Unknown type"*)
 
-  let create typ chunk_size len  =
+  let of_cstruct x =
+    match get_type x with
+    | 0 -> Indirect, x
+    | 1 -> Data, x
+    | _ -> failwith "Unknown type"
+
+  (*let type_of_chunk (idx, _) = idx*)
+
+  let create typ chunk_size len =
     let c = Cstruct.create chunk_size in
     Cstruct.memset c 0x00;
-    set_type c (type_of_int typ);
+    set_type c (int_of_type typ);
     set_length c len;
+    of_cstruct c
+
+  let to_cstruct (_idx, c) =
     c
 
+  (* The Irmin.Contents.Conv signature *)
+  let t = Irmin.Type.like Irmin.Type.cstruct of_cstruct to_cstruct
+  let pp = Irmin.Type.dump t
+  (* TODO: validate *)
+  let of_string s = Result.Ok (of_cstruct @@ Cstruct.of_string s)
```

Comment:
You should probably catch `Failwith _` here and return `Error` in case `s` is malformed.